bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -59,39 +83,57 @@ constructor(
      * Next, it will apply the command to the document, and then add it to the journal itself.
      */
     fun addCommand(command: Journal.Command) {
-        if (!isOpen) {
-            throw IllegalStateException("Cannot add commands to a closed document")
-        }
-        try {
-            command.serialize(journalStream)
-        } catch (ex: BufferOverflowException) {
-            snapshot()
-            command.serialize(journalStream)
+        synchronized(journal) {
+            if (!isOpen) {
+                throw IllegalStateException("Cannot add commands to a closed document")
+            }
+            try {
+                command.serialize(journalStream)
+            } catch (ex: BufferOverflowException) {
+                snapshot()
+                command.serialize(journalStream)
+            }
+            command.apply(document)
+            journal.add(command)
         }
-        command.apply(document)
-        journal.add(command)
     }
 
     /**
      * Save a current document snapshot and clear the journal
      */
     fun snapshot() {
-        if (!isOpen) {
-            throw IllegalStateException("Cannot snapshot a closed document")
+        synchronized(journal) {
+            snapshotInternal()
+        }
+    }
+
+    /**
+     * Save a current snapshot if the journal has passed the high water mark.
+     * Use this in a BG thread to reduce the incidence of snapshotting on the critical path.
+     */
+    fun snapshotIfHighWater() {
+        if (journalStream.bufferSize - journalStream.bytesRemaining() >= highWater) {
+            synchronized(journal) {
+                // Second check in case someone else beat us to it
+                if (journalStream.bufferSize - journalStream.bytesRemaining() >= highWater) {
+                    snapshotInternal()
+                }
+            }
         }
-        JsonHelper.serialize(document, newSnapshotPath)
-        journal.clear()
-        journalStream.clear()
-        journal.serialize(journalStream)
-        newSnapshotPath.renameTo(snapshotPath)
     }
 
     /**
      * Close the journal backing this document. All further modifications will throw exceptions.
      */
     override fun close() {
```

Comment:
This and other public functions could use some more error handling - for instance, if `snapshotInternal()` throws an exception then this would leave the class in an inconsistent state according to the contract of `Closeable`

Reply:
I've added more checks and catches to ensure that every failure is checked, and `close()` completes its task before passing on an exception.